server/src-lib/Hasura/RQL/DDL/Schema/Function.hs

```diff
@@ -188,13 +196,28 @@ newtype TrackFunction
 data FunctionConfig
   = FunctionConfig
   { _fcSessionArgument :: !(Maybe FunctionArgName)
+  -- ^ The argument, if any, to the _tfv2Function to which we'll pass the
+  -- user's session variables during execution. This must be a JSON type
+  -- argument, and disappears from the resulting graphql schema.
+  , _fcAsMutation :: !(Maybe Bool)
+  -- ^ Do we intend this function to end up under the 'mutation' top-level field?
+  --
+  -- We require the 'as_mutation' parameter from users in order to verify their
+  -- intent; since VOLATILE is the default when creating a function, it would
+  -- be easy for users to create a function then track it expecting it to be a
+  -- query, only to find it ended up under the 'mutation' root field.
+  --
+  -- This is Maybe so we can derive instances, with Nothing implying False.
```

Comment:
I don't understand the role of `Maybe` in `_fcAsMutation`. If it's only to derive instances of type classes, then let's obtain those instances in a different way (e.g. spelling them out, if needed).

It would also make sense to use the `RequestedFunction` type I proposed above here, so that we replace a boolean with an informative enumeration of options.

Reply:
> I don't understand the role of Maybe in _fcAsMutation

I considered this and I think the benefits of having consistent To/FromJSON instances outweigh the weirdness. We also are unconcerned with this in other parts of the codebase. I also don't love the situation though. Would be nice if there was a bit more control in aeson, so the serialization format doesn't leak so much into the data models

Reply:
I think in this case it would clarify the Haskell side of things. :-)